src/ringmesh/io/geomodel_adapter_resqml.cpp

```diff
@@ -0,0 +1,677 @@
+                    *rep );
+            }
+        }
+        return true;
+    }
+
+    bool GeoModelAdapterRESQMLImpl::write_volumes()
+    {
+        for( const auto& layer :
+            geomodel_.geol_entities( Layer3D::type_name_static() ) )
+        {
+            AbstractFeatureInterpretation* interp = nullptr;
+            if( layer.has_geological_feature() )
+            {
+                AbstractFeature* feature = find_or_create_feature( layer );
+                AbstractFeatureInterpretation* interp =
+                    find_or_create_interpretation( *feature );
+            }
+            else
+            {
+                Logger::warn( "", layer.gmge(),
+                    " geological feature is empty or not recognized" );
+            }
+
+            // TODO: I do not understand why the interp is not needed...
+            ringmesh_unused( interp );
+
+            std::vector< UnstructuredGridRepresentation* > reps;
+            for( auto i : range( layer.nb_children() ) )
+            {
+                const Region3D& region =
+                    static_cast< const Region3D& >( layer.child( i ) );
+
+                std::string guid( GuidTools::generateUidAsString() );
+                UnstructuredGridRepresentation* rep =
+                    pck_->createUnstructuredGridRepresentation( local_3d_crs_,
+                        guid, region.name(), region.nb_mesh_elements() );
+                reps.push_back( rep );
+
+                std::unique_ptr< double[] > points(
+                    new double[region.nb_vertices() * 3] );
+
+                vec3 p;
+                for( auto v : range( region.nb_vertices() ) )
+                {
+                    p = region.vertex( v );
+                    points[v * 3] = p[0];
+                    points[v * 3 + 1] = p[1];
+                    points[v * 3 + 2] = p[2];
+                }
+
+                std::unique_ptr< ULONG64[] > cumul_faces_cells(
+                    new ULONG64[region.nb_mesh_elements()] );
+
+                std::vector< ULONG64 > face_indices_per_cell;
+                std::vector< ULONG64 > cumul_vertices_face;
+                std::vector< ULONG64 > node_indices_per_face;
+                std::vector< unsigned char > face_righthandness;
+
+                index_t facet_count = 0;
+                for( auto t : range( region.nb_mesh_elements() ) )
+                {
+                    for( auto f : range( region.nb_cell_facets( t ) ) )
+                    {
+                        for( auto v :
+                            range( region.nb_cell_facet_vertices( t, f ) ) )
+                        {
+                            node_indices_per_face.push_back(
+                                region.cell_facet_vertex_index( t, f, v ) );
+                        }
+                        face_indices_per_cell.push_back( facet_count++ );
+
+                        // TODO compute real face righthandness
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1135)